<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [4]:
query = f'''select distinct(employment)
        from vacancies
'''
count = pd.read_sql_query(query, connection)
count

employment
0           Стажировка
1  Частичная занятость
2     Проектная работа
3     Полная занятость

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [5]:
# текст запроса
query_3_1 = f'''select count(*) 
            from vacancies
'''

In [6]:
# результат запроса
count_vac = pd.read_sql_query(query_3_1, connection)
count_vac

count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [7]:
query_emp = f'''select count(*) 
            from employers
'''

In [8]:
# результат запроса
count_emp = pd.read_sql_query(query_emp, connection)
count_emp

count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [9]:
# текст запроса
query_ar = f'''select count(*) 
            from areas
'''

In [10]:
# результат запроса
count_ar = pd.read_sql_query(query_ar, connection)
count_ar

count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [11]:
# текст запроса
query_in = f'''select count(*) 
            from industries
'''

In [12]:
# результат запроса
count_in = pd.read_sql_query(query_in, connection)
count_in

count
0    294

***

#### выводы по предварительному анализу данных:
В базе находится 49197 вакансий 23501 работодателей по 294 сферам деятельности в 1362 регионах.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
query_vac_area = f'''select areas.name, count(vacancies.id) 
            from areas, vacancies where areas.id=vacancies.area_id 
            group by areas.name 
            order by count(vacancies.id) desc
'''

In [14]:
# результат запроса
count_vac_area = pd.read_sql_query(query_vac_area, connection)
count_vac_area

name  count
0             Москва   5333
1    Санкт-Петербург   2851
2              Минск   2112
3        Новосибирск   2006
4             Алматы   1892
..               ...    ...
764       Тарко-Сале      1
765    Новоаннинский      1
766            Бирск      1
767           Сасово      1
768             Анна      1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
query_salary = f'''select count(vacancies.id) 
            from vacancies
            where vacancies.salary_from is not null or vacancies.salary_to is not null
            
'''

In [16]:
# результат запроса
count_salary = pd.read_sql_query(query_salary, connection)
count_salary

count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
query_salary_from = f'''select avg(vacancies.salary_from) 
            from vacancies                    
'''

query_salary_to = f'''select avg(vacancies.salary_to) 
            from vacancies                    
'''

In [18]:
# результат запроса
count_salary_from = pd.read_sql_query(query_salary_from, connection)
count_salary_from
count_salary_to = pd.read_sql_query(query_salary_to, connection)
count_salary_to
print('Нижняя граница зарплаты:',(count_salary_from))
print('Нижняя граница зарплаты:',(count_salary_to))

Нижняя граница зарплаты:             avg
0  71064.657901
Нижняя граница зарплаты:              avg
0  110536.741923


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
# текст запроса
query_sch_em = f'''select count(*) cnt, schedule, employment
            from vacancies 
            group by schedule, employment
            order by cnt desc
'''

In [20]:
# результат запроса
count_sch_em = pd.read_sql_query(query_sch_em, connection)
count_sch_em

cnt          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
# текст запроса
query_ex = f'''select count(*) cnt, experience
            from vacancies 
            group by experience
            order by cnt asc
'''

In [22]:
# результат запроса
count_ex = pd.read_sql_query(query_ex, connection)
count_ex

cnt          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

***

#### выводы по детальному анализу вакансий
В таких регионах как Москва, Санкт-Петербург, Минск, Новосибирск, Алматы представлено самое большое количество вакансий. Средний диапазон зарплат варьируется от 71065 до 110537. Больше всех вакансий с требованием от 1 до 3 лет опыта работы.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [24]:
# текст запроса
query_em_vac = f'''select employers.name, count(vacancies.id) cnt
            from  employers left join vacancies 
            on vacancies.employer_id=employers.id
            group by employers.name
            order by cnt desc
            limit 10
            
'''

In [25]:
# результат запроса
count_em_vac = pd.read_sql_query(query_em_vac, connection)
count_em_vac

name   cnt
0                 Яндекс  1933
1             Ростелеком   491
2               Тинькофф   444
3                   СБЕР   428
4          Газпром нефть   331
5            ИК СИБИНТЕК   327
6                    МТС   292
7                DataArt   247
8  Совкомбанк Технологии   204
9             Первый Бит   176

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
query_em_vac_area = f'''select * from (select areas.name, count(employers.id) cnt_emp, count(vacancies.id) cnt_vac
            from  areas 
            left join employers
            on employers.area=areas.id
            left join vacancies 
            on vacancies.area_id=areas.id 
            group by areas.name
            order by cnt_emp desc            
            ) as foo where cnt_vac=0 limit 1
'''

In [27]:
# результат запроса
count_em_vac_area = pd.read_sql_query(query_em_vac_area, connection)
count_em_vac_area

name  cnt_emp  cnt_vac
0  Россия      410        0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
# текст запроса
query_em_area = f'''select employers.name, count(DISTINCT vacancies.area_id) cnt_areas
            from  employers
            left join vacancies 
            on vacancies.employer_id=employers.id          
            group by employers.name
            order by cnt_areas desc              
            
'''

In [29]:
# результат запроса
count_em_area = pd.read_sql_query(query_em_area, connection)
count_em_area

name  cnt_areas
0                      Яндекс        181
1                  Ростелеком        152
2                  Спецремонт        116
3      Поляков Денис Иванович         88
4                   ООО ЕФИН          71
...                       ...        ...
23170                  СДЕЛКА          0
23171            Alandr Group          0
23172                     СДК          0
23173      Lemon Land Lombard          0
23174                  Махина          0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
# текст запроса
query_em_ind_null = f'''select count(e.id) cnt
            from  EMPLOYERS e
            left outer join EMPLOYERS_INDUSTRIES ei
            on e.id=ei.employer_id
            where ei.employer_id is null
           
'''

In [31]:
# результат запроса
count_em_ind_null = pd.read_sql_query(query_em_ind_null, connection)
count_em_ind_null

cnt
0  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
# текст запроса
query_em_in = f'''select * from(select e.name, count(ei.industry_id) cnt
            from  EMPLOYERS e
            inner join EMPLOYERS_INDUSTRIES ei
            on e.id=ei.employer_id            
           group by e.name
           order by e.name asc) as foo where cnt=4
           
'''

In [33]:
# результат запроса
count_em_in = pd.read_sql_query(query_em_in, connection)
count_em_in

name  cnt
0        101 Интернет    4
1            21vek.by    4
2                2ГИС    4
3                  2К    4
4         4 пикселя +    4
...               ...  ...
1131            ЮРИОН    4
1132  ЮТИП Технологии    4
1133           ЯКласс    4
1134            ЯрНео    4
1135    Яшин&Партнёры    4

[1136 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
# текст запроса
query_em_dev = f'''select * from (select i.name n, count(ei.employer_id) cnt_env
            from  EMPLOYERS_INDUSTRIES ei
            join INDUSTRIES i
            on ei.industry_id = i.id       
           group by i.name, ei.industry_id) as foo where n='Разработка программного обеспечения'
           
'''

In [35]:
# результат запроса
count_em_dev = pd.read_sql_query(query_em_dev, connection)
count_em_dev

n  cnt_env
0  Разработка программного обеспечения     3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
# код для получения списка городов-милионников
import urllib.parse as req
import requests
from bs4 import BeautifulSoup

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)

url
response = requests.get(url).text
page = BeautifulSoup(response, 'html.parser')

links=(page.find_all('table')[1].find_all('a'))
links_new=links[4:]
cities=[]
index=0
for link in links_new:
    if ']' not in link.text:
        cities.append(link.text)        
        index=index+1


print(cities)
l = tuple(cities)


['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград']


In [37]:
# текст запроса
query_cities = f'''SELECT a.name, count(v.id) cnt
    FROM VACANCIES v
    JOIN employers e ON v.employer_id = e.id
    JOIN AREAS a ON v.area_id = a.id
    WHERE e.name ILIKE 'яндекс' and a.name in {l}
    group by a.name  
    UNION ALL
    SELECT 'Total' total, count(v.id)
    FROM VACANCIES v
    JOIN employers e ON v.employer_id = e.id
    JOIN AREAS a ON v.area_id = a.id
    WHERE e.name ILIKE 'яндекс' and a.name in {l}
    order by 2 asc
               
'''


In [38]:
# результат запроса

count_cities = pd.read_sql_query(query_cities, connection)
count_cities

name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

#### выводы по анализу работодателей
Самое большое количество вакансий от компании Яндекс, так же эта компанию лидирует по охвату регионов. 
У 8419 работодателей не указана сфера деятельности.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [40]:
# текст запроса
query_data = f'''select count(*) from VACANCIES where lower(name) like '%data%' or name like '%Data%' or name like '%данн%'or name like '%Данн%'
           
'''

In [41]:
# результат запроса
count_data = pd.read_sql_query(query_data, connection)
count_data

count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [42]:
query_data = f'''     
    select
       count( v.name)
    from vacancies v
    where
        (v.name ~* 'data scien(ce|tist)|исследователь данных|ML|machine learning|машинн.*обучен.*' and v.name !~* 'HTML')
        and (v.name ~* 'junior' or v.experience = 'Нет опыта' or v.employment = 'Стажировка')
'''


In [43]:
# результат запроса
count_data = pd.read_sql_query(query_data, connection)
count_data

count
0     51

In [44]:
query_data = f'''     
    select
        *
    from vacancies v
    where
        (v.name ~* 'data scien(ce|tist)|исследователь данных|ML|machine learning|машинн.*обучен.*' and v.name !~* 'HTML')
        and (v.name ~* 'junior' or v.experience = 'Нет опыта' or v.employment = 'Стажировка')
'''
count_data = pd.read_sql_query(query_data, connection)
count_data

id                                               name  \
0   50533416                            Junior+ CV/ML developer   
1   54825318                           Data Scientist (Junior+)   
2   55553897                     Junior Data Scientist (Стажер)   
3   55529384     Математик, Junior/Middle/Senior Data Scientist   
4   55145641                              Junior Data Scientist   
5   55343709  Data Scientist/Machine Learning Engineer (Deve...   
6   54475492                            Data Scientist (Junior)   
7   55119989                       Intern/Junior Data Scientist   
8   55120211                         Junior Data Scientist (VK)   
9   55120401                       Junior/Middle Data Scientist   
10  55390769          Data Scientist (Computer Vision) - Junior   
11  52158791  Исследователь в области машинного обучения и м...   
12  52396406                             Data Scientist, Junior   
13  55373640  Junior Data Scientist Управления модельных рис...   
14  55552998  Математик-аналитик (data scientist / data anal...   
15  54657714                    Data Scientist (junior+/middle)   
16  55052496                              Junior Data Scientist   
17  50072394                 Research Engineer / Data Scientist   
18  54924841  Стажер направления Машинное обучение и анализ ...   
19  55156731  Стажёр Data Scientist / Группа сервисов Антифр...   
20  54830582                              Junior Data Scientist   
21  50996908          Senior Research Engineer / Data Scientist   
22  55204943                              Ведущий AI/ML инженер   
23  55074464        Data scientist (Junior/Middle) в консалтинг   
24  66142163  Junior Data Scientist (математик-аналитик) уда...   
25  55895592             Junior Data Scientist (Promo Forecast)   
26  55343711  Data Scientist/Machine Learning Engineer (Deve...   
27  55064629                              Data Scientist в Сбер   
28  55343710  Data Scientist/Machine Learning Engineer (Deve...   
29  55127391  Junior Data Scientist (Валидация скоринговых м...   
30  55127390  Junior Data Scientist (Валидация скоринговых м...   
31  55127393  Junior Data Scientist (Валидация скоринговых м...   
32  55127395  Junior Data Scientist (Валидация скоринговых м...   
33  55127396  Junior Data Scientist (Валидация скоринговых м...   
34  55127394  Junior Data Scientist (Валидация скоринговых м...   
35  55127397  Junior Data Scientist (Валидация скоринговых м...   
36  55127398  Junior Data Scientist (Валидация скоринговых м...   
37  55127392  Junior Data Scientist (Валидация скоринговых м...   
38  55095363            Риск - Аналитик (Junior Data scientist)   
39  54702840  Программист Python/ Data Scientist (junior / m...   
40  55104548       Риск-аналитик (ML Engineer / Data Scientist)   
41  54522206            Data Scientist/ML engineer (Python/C++)   
42  54882705       Junior Data scientist/ Младший риск-аналитик   
43  54948266                         Senior Data Scientist (AI)   
44  55046966             Junior Data Scientist for NLP (Python)   
45  54949463          Junior Research Engineer / Data Scientist   
46  54556346                              Ведущий AI/ML инженер   
47  54689300                                     Data Scientist   
48  54945920              Python developer / Junior ML engineer   
49  52396404                             Data Scientist, Junior   
50  55236061          Data Scientist (Computer Vision) / Junior   

                                           key_skills          schedule  \
0           Python\tSQL\tРабота в команде\tOpenCV\tML     Гибкий график   
1   Python\tOpenCV\tTensorflow\tComputer Vision\tН...       Полный день   
2   Python\tSQL\tData Mining\tC++\tМатематический ...     Гибкий график   
3   Python\tMATLAB\tМатематическое моделирование\t...  Удаленная работа   
4   Python\tSQL\tMachine Learning\tData Science\tА...       Полный день   
5   Machine Learning\tDeep Learning\tData Science\...     Гибкий график   


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [45]:
# текст запроса
query_sql = f'''select
                count(*)
                from public.VACANCIES v
                where ((lower(v.name) like '%data scientist%' or
                lower(v.name) like '%data science%' or
                lower(v.name) like '%исследователь данных%' or
                (v.name like '%ML%' and 
                upper(v.name) not like '%HTML%') or
                lower(v.name) like '%machine learning%' or
                lower(v.name) like '%машинн%обучен%')) and
                (v.key_skills ilike '%postgres%' or
                v.key_skills ilike '%SQL%')
'''


In [46]:
# результат запроса
count_sql = pd.read_sql_query(query_sql, connection)
count_sql

count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [47]:
# текст запроса
query_python = f'''     
 select
                count(*)
                from public.VACANCIES v
                where ((lower(v.name) like '%data scientist%' or
                lower(v.name) like '%data science%' or
                lower(v.name) like '%исследователь данных%' or
                (v.name like '%ML%' and 
                upper(v.name) not like '%HTML%') or
                lower(v.name) like '%machine learning%' or
                lower(v.name) like '%машинн%обучен%')) and
                (v.key_skills ilike '%Python%')
        ''' 


In [48]:
# результат запроса
count_python = pd.read_sql_query(query_python, connection)
count_python

count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [49]:
# текст запроса
query_skills = f'''     
select round(avg(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1), 2) as skills_avg
    FROM public.vacancies
    where key_skills is not null and (name ilike '%data scientist%' or
        name ilike '%data science%' or
        name ilike '%исследователь данных%' or
        name ilike '%machine learning%' or
        name ilike '%машинн%обучен%' or
        (name not like '%HTML%' and name like '%ML%'))
        ''' 

In [50]:
# результат запроса
count_skills = pd.read_sql_query(query_skills, connection)
count_skills

skills_avg
0        6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [51]:
# текст запроса
query_sal= f''' select distinct v.experience,
                round(AVG(coalesce((salary_from+salary_to)/2, salary_from, salary_to)),0) 
                from public.VACANCIES v
                where ((salary_from is not null or
                salary_to is not null)) and
                ((v.name ilike '%data scientist%' or
                v.name ilike '%data science%' or
                v.name ilike '%исследователь данных%' or
                v.name ilike '%machine learning%' or
                v.name ilike '%машинн%обучен%') or
                (v.name like '%ML%' and 
                v.name not like '%HTML%'))
                group by 1
'''

In [52]:
# результат запроса
count_sal = pd.read_sql_query(query_sal, connection)
count_sal

experience     round
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

In [53]:
connection.close()

***

#### выводы по предметному анализу
51 вакансия предназначена для начинающих специалистов в DS. В 351 вакансиях Python указан как ключевой навык, в 201 - sql и postgres. Средняя зп для начинающего специалиста без опыта - около 75 тыс. руб., а специалисты с опытом работы от 3 до 6 лет могут претендовать на зп в среднем 243 тыс. руб.

# Общий вывод по проекту

Среди всех вакансий большее число размещено в городах Москва, Санкт-Петербург, Минск. Основной график работы - полная занятость на полный день. Среди них только 51 вакансия для начинающих специалистов в DS. 